src/utils/index.ts

```diff
@@ -0,0 +1,105 @@
+import { BigNumber } from '@ethersproject/bignumber'
+import { abi as IUniswapV2Router02ABI } from '@uniswap/v2-periphery/build/IUniswapV2Router02.json'
+import { ChainId, JSBI, Percent, Token, CurrencyAmount, Currency, ETHER } from '@pancakeswap/sdk'
+import { ROUTER_ADDRESS } from '../config/constants'
+import { BASE_BSC_SCAN_URLS } from '../config'
+import { TokenAddressMap } from '../state/lists/hooks'
+
+// returns the checksummed address if the address is valid, otherwise returns false
+export function isAddress(value: any): string | false {
+  try {
+    return getAddress(value)
+  } catch {
+    return false
+  }
+}
+
+export function getBscScanLink(
+  data: string | number,
+  type: 'transaction' | 'token' | 'address' | 'block' | 'countdown',
+  chainId: ChainId = ChainId.MAINNET,
+): string {
+  switch (type) {
+    case 'transaction': {
+      return `${BASE_BSC_SCAN_URLS[chainId]}/tx/${data}`
+    }
+    case 'token': {
+      return `${BASE_BSC_SCAN_URLS[chainId]}/token/${data}`
+    }
+    case 'block': {
+      return `${BASE_BSC_SCAN_URLS[chainId]}/block/${data}`
+    }
+    case 'countdown': {
+      return `${BASE_BSC_SCAN_URLS[chainId]}/block/countdown/${data}`
+    }
+    default: {
+      return `${BASE_BSC_SCAN_URLS[chainId]}/address/${data}`
+    }
+  }
+}
+
+// shorten the checksummed version of the input address to have 0x + 4 characters at start and end
+export function shortenAddress(address: string, chars = 4): string {
+  const parsed = isAddress(address)
+  if (!parsed) {
+    throw Error(`Invalid 'address' parameter '${address}'.`)
+  }
+  return `${parsed.substring(0, chars + 2)}...${parsed.substring(42 - chars)}`
+}
+
+// add 10%
+export function calculateGasMargin(value: BigNumber): BigNumber {
+  return value.mul(BigNumber.from(10000).add(BigNumber.from(1000))).div(BigNumber.from(10000))
+}
+
+// converts a basis points value to a sdk percent
+export function basisPointsToPercent(num: number): Percent {
+  return new Percent(JSBI.BigInt(num), JSBI.BigInt(10000))
+}
+
+export function calculateSlippageAmount(value: CurrencyAmount, slippage: number): [JSBI, JSBI] {
+  if (slippage < 0 || slippage > 10000) {
+    throw Error(`Unexpected slippage value: ${slippage}`)
+  }
+  return [
+    JSBI.divide(JSBI.multiply(value.raw, JSBI.BigInt(10000 - slippage)), JSBI.BigInt(10000)),
+    JSBI.divide(JSBI.multiply(value.raw, JSBI.BigInt(10000 + slippage)), JSBI.BigInt(10000)),
+  ]
+}
+
+// account is not optional
+export function getSigner(library: Web3Provider, account: string): JsonRpcSigner {
+  return library.getSigner(account).connectUnchecked()
+}
+
+// account is optional
+export function getProviderOrSigner(library: Web3Provider, account?: string): Web3Provider | JsonRpcSigner {
+  return account ? getSigner(library, account) : library
+}
+
+// account is optional
+export function getContract(address: string, ABI: any, library: Web3Provider, account?: string): Contract {
```

Comment:
Effectively a duplicate of https://github.com/pancakeswap/pancake-frontend/blob/7b7c12cc84a98ebfba2b0e3b43502b2a712c3033/src/utils/contractHelpers.ts#L56?